language/resource-api/README.md

```diff
@@ -356,6 +356,43 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
 ```
 
 
+### Provider feature: `custom_insync`
+
+Allows for overriding default property `insync?` checking using custom logic in the provider.
+
+```ruby
+# lib/puppet/type/apt_key.rb
+Puppet::ResourceApi.register_type(
+  name: 'apt_key',
+  features: [ 'custom_insync' ],
+)
+
+# lib/puppet/provider/apt_key/apt_key.rb
+class Puppet::Provider::AptKey::AptKey
+  def insync?(context, name, attribute_name, is_hash, should_hash)
+    case attribute_name
+    when :some_array_attribute
+      if should_hash[:force]
+        context.notice("Checking an order independent array")
+        # This will return true if and only if the members of both
+        # arrays match entirely, regardless of order.
+        return is_hash[attribute_name].sort == should_hash[attribute_name].sort
+      else
+        context.notice("Checking a subset match array")
+        # This will return true so long as all values in the array for
+        # the should_hash are also in the array of the is_hash.
+        return (is_hash[attribute_name] & should_hash[attribute_name]).sort == should_hash[attribute_name].sort
+      end
+    end
+  end
+```
+
+Specifying the `custom_insync` feature flag ensures that all of the property attributes on the type will call the `insync?` method in the provider, if it exists, enabling custom state comparisons as complex or simple as required. The `insync?` method in the provider has access to the `context` for logging, the name of the resource being checked, the attribute that is currently being inspected for insync status, and hashes representing the current state of the resource as well as the canonicalized should state of the resource.
```

Comment:
```suggestion
When the `custom_insync` feature flag is specified, `insync?` is called once for each of the properties. This can be used to implement custom state comparisons, when the default value-equality is not sufficient. The `insync?` method receives a `context` for logging, the `name` of the resource being checked, the `property_name` that is currently being inspected for insync status, the current state of the resource in `is_hash`, as well as the desired state (canonicalized) in `should_hash`.
```